R/facet-.r

```diff
@@ -1,62 +1,190 @@
-#' Facet specification.
-#'
-#' Create new facetting specification.  For internal use only.
-#'
-#' @param ... object fields
-#' @param shrink shrink scales to fit output of statistics, not raw data
-#' @keywords internal
-#' @export
-facet <- function(..., shrink = TRUE, subclass = c()) {
-  structure(list(..., shrink = shrink), class = c(subclass, "facet"))
-}
+#' @include ggproto.r
+NULL
 
 #' Is this object a facetting specification?
 #'
 #' @param x object to test
 #' @keywords internal
 #' @export
-is.facet <- function(x) inherits(x, "facet")
+is.facet <- function(x) inherits(x, "Facet")
 
+Facet <- ggproto("Facet", NULL,
+  shrink = FALSE,
+  params = list(),
 
-# Figure out layout from data from plot and all layers.
-#
-# This creates the layout data frame which maps from data values to
-# panel coordinates: ROW, COL and PANEL. It also records the panels that
-# contribute to each x and y scale.
-#
-# @param data a list of data frames (one for the plot and one for each
-#   layer)
-facet_train_layout <- function(facet, data)
-  UseMethod("facet_train_layout")
 
-facet_map_layout <- function(facet, data, layout)
-  UseMethod("facet_map_layout")
+# Layout interface --------------------------------------------------------
 
-facet_render <- function(facet, panels_grob, coord, theme, geom_grobs)
-  UseMethod("facet_render")
+  train = function(self, data) {
+    self$compute_layout(data, self$params)
+  },
+  map = function(self, data, layout) {
+    self$map_data(data, layout, self$params)
+  },
+  render_below = function(self, data, layout, x_scales, y_scales) {
```

Comment:
Ok, lets change it — back and front are what drawing programs use, e.g. "send to back"